design/baremetal-operator/raid-api.md

```diff
@@ -0,0 +1,198 @@
+need one drive with minimal configuration, many classes of server
+hardware need to have RAID storage configured to present fast and
+reliable storage to the OS. While it is possible today to
+pre-configure the hosts, it would be more convenient for users is
+metal3 could do that work for them.
+
+### Goals
+
+1. Provide an API to specify the RAID storage configuration to use
+   when provisioning a host.
+2. Outline the controller/provisioner logic changes needed to pass the
+   instructions to Ironic to have the storage configured.
+3. Support basic RAID levels `0`, `1`, and `1+0`.
+
+### Non-Goals
+
+1. Provide default RAID configuration.
+2. Retain existing RAID configuration on a host if it matches the
+   desired configuration at the time of provisioning (we make no
+   assumptions about preserving the state of the storage system or its
+   contents).
+3. Re-provision a host if the RAID settings change. (Users are
+   expected to explicitly deprovision the host and then reprovision
+   it.)
+4. Support every possible RAID configuration.
+
+## Proposal
+
+### User Stories
+
+1. As a user, I can specify the RAID configuration for a host before
+   provisioning it.
+2. As a user, I can choose between hardware and software RAID for each
+   host.
+
+## Design Details
+
+### API Changes
+
+The `BareMetalHost` API will change to accept RAID configuration
+inputs and to reflect the settings that were used when a host is
+provisioned (in case they are changed after provisioning).
+
+New `spec` fields will be added in a section called `raid`, with
+separate fields describing the hardware and software RAID settings. If
+the hardware settings are provided, the software settings are ignored.
+
+For hardware RAID, the inputs describe how to combine physical disks
+into a RAID volume via the RAID controller hardware in the host.
+
+```yaml
+spec:
+  raid:
+    hardwareVolumes:
+      - sizeGibibytes: 1000
+        level: 1
+        name: volume-name
+        rotational: true
+        numberOfPhysicalDisks: 2
+```
+
+- The RAID level for a volume is required and should be expressed as
+  one of `"1"`, `"0"`, or `"1+0"`.
+- The size for a volume is optional and should be expressed in
+  gibibytes.
+- The number of disks to use for the volume defaults to a minimum
+  value based on the RAID level.
+
+For software RAID, the inputs describe how to configure the operating
+system to see physical disks in a RAID configuration through software
+drivers, without using any special hardware.
+
+```yaml
+spec:
+  raid:
+    softwareVolumes:
+      - sizeGibibytes: 1000
+        level: 1
+        physicalDisks:
+          - deviceName: /dev/sda
+          - deviceName: /dev/sdb
+```
+
+- The RAID level for a volume is required and should be expressed as
+  one of `"1"`, `"0"`, or `"1+0"`.
+- The size for a volume is optional and should be expressed in
+  gibibytes.
+- The disks to use are specified using the same syntax as the
+  `rootDeviceHints`.
```

Comment:
This is true, but for now deviceName is the only hint supported.

Reply:
This looks like a great starting point. Unless we have a dire need for supporting many other device hints? Do we have some use cases in mind for those?

Reply:
Is the use of names a limitation of Ironic or of the implementation in the baremetal-operator? Perhaps we should just change `physicalDisks` to be a list of device names?

Reply:
I implemented this in `baremetal-operator`, and [modify gophercloud](https://github.com/gophercloud/gophercloud/pull/1982) to support `physicalDisks` in software raid.

Reply:
> Is the use of names a limitation of Ironic or of the implementation in the baremetal-operator? Perhaps we should just change `physicalDisks` to be a list of device names?

I'm fine with replacement with simply disk names, but this discussion is about specific mention of the _root device hints_, which for now, is simply the name of the device, other parameters perhaps could be very particular details like models, vendor names, rotational (boolean) and so on. The whole list is here
https://docs.openstack.org/ironic/pike/install/include/root-device-hints.html

I was just wondering if it's okay to only have a ``deviceName`` as the root device hint. Which I think is sufficient for most use cases. And unless we can figure out another use case where people would like to choose their root devices based on vendor names, or SCSI bus addresses etc?

Reply:
No, my comment was not about root device hints, it's about the selection of disks for RAID.
I missed that @Hellcatlk has implemented that other selection criteria in gophercloud, so IIUC this description is now accurate and my comment is obsolete.

Reply:
I see. That's alright then, no need to discuss the hints.